File: Deal/src/main/java/ru/leonov/deal/exceptions/ApplicationException.java
package ru.leonov.deal.exceptions;

import lombok.EqualsAndHashCode;
import lombok.Getter;

/**
 * Exception that is thrown during loan offer application process.
 */
@EqualsAndHashCode(callSuper = false)
public class ApplicationException extends RuntimeException {

    @Getter
    private final ExceptionCause exceptionCause;

    public ApplicationException(ExceptionCause exceptionCause) {
        super(exceptionCause.getUserFriendlyMessage());
        this.exceptionCause = exceptionCause;
    }

    /**
     * Exception cause enum.
     */
    public enum ExceptionCause {

        APPLICATION_NOT_FOUND("Requested application not found in database.");

        @Getter
        private final String userFriendlyMessage;

        ExceptionCause(String userFriendlyMessage) {
            this.userFriendlyMessage = userFriendlyMessage;
        }

    }
}

File: Deal/deal_impl/src/main/kotlin/ru/leonov/deal/exception/ApplicationException.kt
package ru.leonov.deal.exception

/**
 * Exception that is thrown during loan offer application process.
 */
data class ApplicationException(val exceptionCause: ExceptionCause) :
    RuntimeException(exceptionCause.userFriendlyMessage) {

    /**
     * Exception cause enum.
     */
    enum class ExceptionCause(val userFriendlyMessage: String) {
        APPLICATION_NOT_FOUND("Requested application not found in database.")
    }

}
